### DISCUSSION TITLE
Should there be a way to turn off the playground web ui in production environments?

### DISCUSSION THREAD

**Discussion by cvignola** [ClickHouse/ClickHouse#30339] [Q&A] [answered]
Some, possibly most, production configurations would disallow use of the playground web ui.  For such configurations, its  availability represents an additional security attack vector.  Is there a defined way to configure it off?  I found no evidence that there is.  Therefore, would it be reasonable to introduce a configuration option to disable that server path?

**Comment by alexey-milovidov** [comment]
Playground web UI does not introduce any security risks, because it is a static HTML page.
When you do queries in Play UI, it is sending the queries to the HTTP interface, that you can limit in config.xml and users.xml.
Play UI does not open anything by itself, instead it is simply reusing the HTTP interface.

So, if HTTP interface is not available to adversaries, then Play UI will also be unavailable.
See `listen` and `http_port` in `config.xml` and `networks` in `users.xml`.
By default it's only available from localhost.

**Reply by ramazanpolat** [reply]
Although it is just a static HTML file which uses the `http_port`, the play interface makes it look like ClickHouse **is serving** a web app. This will make it hard to convince people that it does not. So being able to disable it won't hurt. It may save hours spent to convice security guys.

**Comment by ramazanpolat** [comment]
@cvignola you can override it by adding following lines to `config.xml`:
```xml
<http_handlers>
<rule> 
  <url>/play</url> 
  <methods>GET</methods> 
  <handler> 
    <type>static</type> 
    <status>403</status> 
    <content_type>text/plain; charset=UTF-8</content_type> 
    <response_content></response_content> 
  </handler> 
</rule>
</http_handlers>
```

**Reply by bkuschel** [reply]
The challenge is that as soon as you define a handler, it removes all the default ones, so they have to be added to the config manually.

**Reply by bkuschel** [reply]
though i found a `defaults` that might do the trick

**Reply by ramazanpolat** [reply]
can you show me how you did that trick?

**Comment by bkuschel** [comment, marked as answer]
@ramazanpolat this did it...

```xml
    <http_handlers>
        <rule>
            <url>/play</url>
            <methods>GET</methods>
            <handler>
                <type>static</type>
                <status>403</status>
                <content_type>text/plain; charset=UTF-8</content_type>
                <response_content></response_content>
            </handler>
        </rule>
        <defaults/>
    </http_handlers>
```

**Reply by ramazanpolat** [reply]
I belive it is not documented and I'm curious about how you found it :)

**Reply by bkuschel** [reply]
code review :)

**Comment by bkuschel** [comment]
@alexey-milovidov i think this can be closed since there is indeed a away to disable play in configuration 🎉

**Comment by Hiieu** [comment]
You can still able to access the `/play` endpoint by going to this endpoint `/play/randomstring`.
Unless, there's an option to catch any endpoints with a wildcard, then I don't think it's possible to disable the playground.

**Reply by DecLanRonin** [reply]
<url>regex:^/play.*</url>

**Comment by maffelbaffel** [comment]
For anyone looking for a copy paste. This works for me:
```
spec:
    configuration:
      files:
        config.d/disable_play.xml: |
          <clickhouse>
            <http_handlers>
                <rule>
                    <url>regex:^/play.*</url>
                    <methods>GET</methods>
                    <handler>
                        <type>static</type>
                        <status>403</status>
                        <content_type>text/plain; charset=UTF-8</content_type>
                        <response_content></response_content>
                    </handler>
                </rule>
                <defaults/>
            </http_handlers>
          </clickhouse>
```
